server/aesdsocket.c

```diff
@@ -14,140 +14,288 @@
 #include<string.h>
 #include<arpa/inet.h>
 #include<netinet/in.h>
+#include "queue.h"
+#include <pthread.h>
+#include <time.h>
+#include <sys/time.h>
+#include <string.h>
 
-// struct sockaddr{
-//     sa_family_t sa_family;
-//     char sa_data[14];
-// }
+//pthread_mutex_t *mutex;
 
 #define PORT "9000"
-#define BACKLOG (6)
+#define BACKLOG (5)
 #define BUFFER_SIZE (1024)
 #define FILE_PATH "/var/tmp/aesdsocketdata"
 
+pthread_mutex_t mutex;
+int sockfd,new_sockfd,file_fd;
+bool interrupted=false;
+char* buf=NULL;
+struct addrinfo *server_info=NULL;
+int total_connections=0;
+bool timer_alarm=false;
+
+struct thread_data{
+    pthread_t thread_id;
+    int socketfd;
+    struct sockaddr *client;
+    bool connection_complete_success;
+    SLIST_ENTRY(thread_data) entries;
+};
+
+SLIST_HEAD(slisthead,thread_data) head=SLIST_HEAD_INITIALIZER(&head);
+
+int set_time(){
+    int ret;
+    struct itimerval timer_1;
+    getitimer(ITIMER_REAL,&timer_1);
+    timer_1.it_value.tv_sec=10;
+    timer_1.it_value.tv_usec=0;
+    timer_1.it_interval.tv_sec=10;
+    timer_1.it_interval.tv_usec=0;
+    ret=setitimer(ITIMER_REAL,&timer_1,NULL);
+    return(ret);
+}
+int reset_time(){
+    int ret;
+    struct itimerval timer_1;
+    getitimer(ITIMER_REAL,&timer_1);
+    timer_1.it_value.tv_sec=0;
+    timer_1.it_value.tv_usec=0;
+    timer_1.it_interval.tv_sec=0;
+    timer_1.it_interval.tv_usec=0;
+    ret=setitimer(ITIMER_REAL,&timer_1,NULL);
+    return(ret);
+}
+
+int write_time_to_file(int fd){
+    char outstr[100]={};
+    char buffer[100];
+    time_t t;
+    struct tm *tmp;
+    const char* fmt= "%a, %d %b %Y %T %z";
+    t = time(NULL);
+    tmp = localtime(&t);
+    if (tmp == NULL) {
+        perror("localtime");
+        return(-1);
+    }
+
+    if (strftime(outstr, sizeof(outstr), fmt, tmp) == 0) {
+        return(-1);
+    }
+    strcpy(buffer,"timestamp:");
+    strcat(buffer,outstr);
+    strcat(buffer,"\n");
+    write(fd,buffer,strlen(buffer)*sizeof(char));
+    return(0);
+}
+
+
+void* thread_function(void* thread_param){
+    int nr,ret;
+    int size_recived=0,size_read=0,s_recv,s_send,current_size=0;
+    char clientIP[INET6_ADDRSTRLEN];
+    struct thread_data* thread_info = (struct thread_data *)thread_param;
+    inet_ntop(AF_INET,thread_info->client,clientIP,sizeof(clientIP));
+    bool recv_complete=false;
+    buf=malloc(BUFFER_SIZE*sizeof(char*));
+    while(!recv_complete){
+        buf=(char*)(realloc(buf,current_size+BUFFER_SIZE));
+        s_recv=recv(thread_info->socketfd,(buf+current_size),BUFFER_SIZE,0);
+        if(s_recv==-1){
+            syslog(LOG_USER, "Error while recieving");
+            perror("recv");
+            exit(-1);
+        }
+        size_recived=current_size+s_recv;
+        if(size_recived>0 && buf[size_recived-1]=='\n'){
```

Comment:
@Sona-Subair will this handle the case where multiple packets are sent in a buffer or the buffer does not end with `\n`?  See related comment at https://github.com/cu-ecen-aeld/assignments-3-and-later-Sona-Subair/pull/3#discussion_r988512074